### ISSUE TITLE
If I wanna to add a validator to a living net in tendermint with four validators, What should I do

### ISSUE THREAD

**Issue by Quieting** [tendermint/tendermint#1291]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
My ABCI app base on PersistentDummyApplication.

**Merkleeyes version** (use `git rev-parse --verify HEAD`, skip if you don't use it):

**Environment**:
- **macOS** (e.g. from /etc/os-release):
- **Install tools**:glid
- **Others**:tendermint 0.16.0

**What happened**:

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

**Anything else do we need to know**:
I've done a test: 
In EndBlock function, I add a new validator to ResponseEndBlock.ValidatorUpdates and return it. When I send message by "http://192.168.1.2:46657/broadcast_tx_commit",the new validator don't join net. The Result Validators includes the new validator by "http://192.168.1.2:46657/dump_consensus_state".But, other validators don't.
What‘s wrong or what to miss?
Thanks!
Forgive, my english is bad.

**Comment by melekes** [comment]
Is there any information in the logs?

**Comment by Quieting** [comment]
It's my steps:
1. run "tendermint init" on four difference computers

2. copy "genesis.json".pub_key from each computer. This is final genesis.json：

`{"genesis_time":"0001-01-01T00:00:00Z","chain_id":"test-chain-Kpb4Zf","validators":[{"pub_key":{"type":"ed25519","data":"5B73D37C087789D4331A7D4A7D3865FED584FABA06B272693E0A71A00F18FF04"},"power":10,"name":""},{"pub_key":{"type":"ed25519","data":"68AEA41174FA61FADC89599BDD3F38DD2A45F502CE70374FF348504C8C9A5C13"},"power":10,"name":""},{"pub_key":{"type":"ed25519","data":"F43FA06A9FB752690A15B412844EC14DD19C0B47A9A39BB5C23E48814DC04865"},"power":10,"name":""},{"pub_key":{"type":"ed25519","data":"398B78D685F80E18E011A514060136C092D983C4CF5D6B5E78389E3F2AF4ECB6"},"power":10,"name":""}],"app_hash":""}`

everyone is same.

3. run `tendermint node 
 --p2p.persistent_peers=192.168.1.3:46656,192.168.1.4:46656,192.168.1.5:46656\
 --home /root/.tendermint --log_level="*:error"` on  `192.168.1.2` computer.

4.On a new computer, run `tendermint init`, then update "genesis.json".chain_id"  as the same as  "test-chain-Kpb4Zf", run `tendermint node --p2p.persistent_peers=192.168.1.2:46656 `.Found some errors:
`Error attempting to add vote                 module=consensus err="vote.ValidatorAddress (31334230343834313143313530364136384246303443373144444646433835393139444634464343) does not match address (8F072B969F7982ED764C50BB30D1F29F64D0DAE4) for vote.ValidatorIndex (0)\nEnsure the genesis file is correct across all validators.: Invalid validator address"`

`E[03-09|02:56:43.457] Error attempting to add vote                 module=consensus err="Cannot find validator 1 in valSet of size 1: Invalid validator index"`
`E[03-09|02:56:43.457] Error attempting to add vote                 module=consensus err="Cannot find validator 2 in valSet of size 1: Invalid validator index"`
`E[03-09|02:56:43.458] Error attempting to add vote                 module=consensus err="Cannot find validator 3 in valSet of size 1: Invalid validator index"`

**Comment by Quieting** [comment]
@melekes thanks

**Comment by Quieting** [comment]
The problem was solved via "have the same genesis file everywhere".Thanks for your suggest.
But new doubts appeared. When a node as a non-validator joins tendermint network,then request `/broadcast_tx_commit ` become a validator. I found id via request `/dump_consensus_state`. If the new validator `voting_power` is big(more than total `voting_power`), the network stop make blocks and print 
`I[03-09|08:54:24.565] Ensure peers   module=p2p numOutPeers=1 numDialing=0 numToDial=9`
`I[03-09|08:54:24.565] We need more addresses. Sending pexRequest to random peer module=p2p peer="Peer{MConn{192.168.1.2:46656} dacac02e804a55a5d60e3784a1555efb81d3cac5 out}"`.
Are right?

**Comment by melekes** [comment]
I don't see any errors. Can you attach the full log and output of `/dump_consensus_state `?

**Comment by Quieting** [comment]
@melekes I'm sorry to get back to you since last time. I got entangled in something during this period. I have to suspend  of learning tendermint. Thanks for your help very much.

**Comment by EnchanterIO** [comment]
I read the whole documentation of ethermint/tendermint today/yesterday and commit,begin and endblock is the only parts I skipped... damn murphy laws.

@melekes Checking the code of tendermint (v.0.14 as I use ethermint for app logic):

`state/state.go::SetBlockAndValidators`
It adds a new validator to the ValidatorSet and then it updates the local Mempool using ApplicationLogic. Where does the validation of "new validator" happens? There should be some ConsensousEngine logic executed right?

Is it happening asynchronously later? I think each transaction/update of a Mempool is broadcasted to the whole network by a ProposerNode and then the ConcensusRound will validate if they will accept the validator? Otherwise one node could just add a lot of fake/malicious validators or so.

**Comment by melekes** [comment]
> Where does the validation of "new validator" happens? There should be some ConsensousEngine logic executed right?

there is no logic except Tendermint making changes to validator set according to the EndBlock https://github.com/tendermint/tendermint/blob/37385cb1d4155bd23ec1f810739f531e8e92b0fb/state/execution.go#L318.

> Otherwise one node could just add a lot of fake/malicious validators or so.

but others won't do the same. end blocks are not being propagated like transactions. If app A changes validator set, B can choose not to (unless B is malicious too). Then B along with other honest validators will ignore A because they will require 2/3 of the current val set to sign anything.

You also may be interested in this change https://github.com/tendermint/tendermint/pull/1756

Also, we're working on a new ethermint, which is in proof of concept stage right now https://github.com/ledgerwatch/ethermint.

**Comment by EnchanterIO** [comment]
> Then B along with other honest validators will ignore A because they will require 2/3 of the current val set to sign anything.

Hm, Let's say we have 4 nodes: A,B,C,D. A is malicious.

If malicious A would add 5 more validators, specifying high voting power, these changes **won't** be propagated to the network therefore B,C,D **won't** be connected to these 5 extra malicious validators. Therefore the voting power of A is still just the original value the B,C,D accepted when peered up as was specified in genesis?

So the network would "split" and A could do malicious transactions only between its own 5 malicious validators which is useless.

Can u confirm this or did I missunderstand? @melekes 

Thx

**Comment by melekes** [comment]
> Therefore the voting power of A is still just the original value the B,C,D accepted when peered up as was specified in genesis?

Correct.

**Comment by diddyxking** [comment]
Does this still hold true by using elections? I have the same error after I have added a validator node by using an election. 4 Nodes with the same genesis.json and config.toml were set up in advance. The 5th node only has a matching chain_id in genesis.json and all nodes in the persisent_peers field in the config.toml. All nodes connected successfully (even without an election) but data is not synced between those 4 nodes and the 5th node. I have tried all combinations but I can't get elections to work and I always get the same error as the op.
